### ISSUE TITLE
ProstT5 issue: iteration over a 0-d array

### ISSUE THREAD

**Issue by vdruelle** [gbouras13/phold#47]
* phold version: 0.1.4
* Python version: 3.11.9
* Operating System: Ubuntu 20.04.6 LTS

### Description
I'm trying to generate annotation for a couple of phages from the BASEL collection  (like this one from NCBI https://www.ncbi.nlm.nih.gov/nuccore/2071745857) to test the performance of the tool. I am first generating a genbank file using Pharokka, which seems to be fine since the tool completes the job and I obtain the pharokka.gbk file in the output folder.

I then try to use phold on this file with the command:
`phold run -i output/EM60_pharokka/pharokka.gbk -o output/EM60_phold -t 8 -f --cpu`
I'm running the cpu version since my GPU doesn't have enough memory for the gpu version.

The tool starts but eventually fails at the ProstT5 prediction step. I'm copy pasting the output of the terminal below. I tried figuring out what was the problem but it was unconclusive. 

Do you have an idea where the issue comes from ?
Thanks and have a great day !

```
2024-07-02 14:35:52.978 | INFO     | phold.utils.validation:instantiate_dirs:70 - Checking the output directory output/EM60_phold
2024-07-02 14:35:52.979 | INFO     | phold.utils.validation:instantiate_dirs:76 - --force was specified even though the output directory does not already exist. Continuing


.______    __    __    ______    __       _______  
|   _  \  |  |  |  |  /  __  \  |  |     |       \ 
|  |_)  | |  |__|  | |  |  |  | |  |     |  .--.  |
|   ___/  |   __   | |  |  |  | |  |     |  |  |  |
|  |      |  |  |  | |  `--'  | |  `----.|  '--'  |
| _|      |__|  |__|  \______/  |_______||_______/ 
                                                   


2024-07-02 14:35:52.988 | INFO     | phold.utils.util:begin_phold:72 - phold: annotating phage genomes with protein structures
2024-07-02 14:35:52.988 | INFO     | phold.utils.util:begin_phold:74 - You are using phold version 0.1.4
2024-07-02 14:35:52.988 | INFO     | phold.utils.util:begin_phold:75 - Repository homepage is https://github.com/gbouras13/phold
2024-07-02 14:35:52.988 | INFO     | phold.utils.util:begin_phold:76 - You are running phold run
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:77 - Listing parameters
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --input output/EM60_pharokka/pharokka.gbk
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --output output/EM60_phold
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --threads 8
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --force True
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --prefix phold
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --evalue 0.001
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --database None
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --batch_size 1
2024-07-02 14:35:52.989 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --sensitivity 9.5
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --keep_tmp_files False
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --cpu True
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --omit_probs False
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --finetune False
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --finetune_path None
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --split False
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --split_threshold 60.0
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --card_vfdb_evalue 1e-10
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --separate False
2024-07-02 14:35:52.990 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --max_seqs 1000
2024-07-02 14:35:52.993 | INFO     | phold.utils.validation:check_dependencies:117 - Foldseek version found is v8.ef4e960
2024-07-02 14:35:52.994 | INFO     | phold.utils.validation:check_dependencies:126 - Foldseek version is ok
2024-07-02 14:35:52.994 | INFO     | phold.databases.db:validate_db:234 - Checking Phold database installation in /home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/database
2024-07-02 14:35:52.994 | INFO     | phold.databases.db:validate_db:237 - All Phold databases files are present
2024-07-02 14:35:52.995 | INFO     | phold.io.handle_genbank:get_genbank:57 - Checking if input output/EM60_pharokka/pharokka.gbk is a Genbank file
2024-07-02 14:35:53.007 | INFO     | phold.utils.validation:validate_input:50 - Successfully parsed input output/EM60_pharokka/pharokka.gbk as a Genbank format file
2024-07-02 14:35:53.008 | INFO     | phold.features.predict_3Di:get_T5_model:121 - Using device: cpu
2024-07-02 14:35:53.008 | INFO     | phold.features.predict_3Di:get_T5_model:127 - Loading T5 from: /home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/database/Rostlab/ProstT5_fp16
2024-07-02 14:35:53.009 | INFO     | phold.features.predict_3Di:get_T5_model:128 - If /home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/database/Rostlab/ProstT5_fp16 is not found, it will be downloaded
You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
2024-07-02 14:35:54.396 | INFO     | phold.features.predict_3Di:get_T5_model:138 - Rostlab/ProstT5_fp16 loaded
2024-07-02 14:35:54.403 | INFO     | phold.features.predict_3Di:get_embeddings:362 - Beginning ProstT5 predictions
2024-07-02 14:35:54.403 | INFO     | phold.features.predict_3Di:get_embeddings:369 - Using models in full-precision
2024-07-02 14:40:08.983 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0058 MZ501093.1 prediction has length 0
2024-07-02 14:43:19.564 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0166 MZ501093.1 prediction has length 0
2024-07-02 14:43:21.652 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0174 MZ501093.1 prediction has length 0
2024-07-02 14:43:28.210 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0192 MZ501093.1 prediction has length 0
2024-07-02 14:43:31.288 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0200 MZ501093.1 prediction has length 0
2024-07-02 14:43:36.685 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0215 MZ501093.1 prediction has length 0
2024-07-02 14:43:39.533 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0221 MZ501093.1 prediction has length 0
Traceback (most recent call last):
  File "/home/valentin/miniconda3/envs/pholdENV/bin/phold", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/__init__.py", line 1355, in main
    main_cli()
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/__init__.py", line 281, in run
    subcommand_predict(
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/subcommands/predict.py", line 125, in subcommand_predict
    prediction_success = get_embeddings(
                         ^^^^^^^^^^^^^^^
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/features/predict_3Di.py", line 526, in get_embeddings
    write_predictions(predictions, output_3di, proteins_flag)
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/features/predict_3Di.py", line 206, in write_predictions
    [
  File "/home/valentin/miniconda3/envs/pholdENV/lib/python3.11/site-packages/phold/features/predict_3Di.py", line 210, in <listcomp>
    list(map(lambda yhat: ss_mapping[int(yhat)], yhats))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: iteration over a 0-d array
```

**Comment by gbouras13** [comment]
Hi @vdruelle ,

The error seems to be caused by the warning lines:

```
2024-07-02 14:40:08.983 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0058 MZ501093.1 prediction has length 0
2024-07-02 14:43:19.564 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0166 MZ501093.1 prediction has length 0
2024-07-02 14:43:21.652 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0174 MZ501093.1 prediction has length 0
2024-07-02 14:43:28.210 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0192 MZ501093.1 prediction has length 0
2024-07-02 14:43:31.288 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0200 MZ501093.1 prediction has length 0
2024-07-02 14:43:36.685 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0215 MZ501093.1 prediction has length 0
2024-07-02 14:43:39.533 | WARNING  | phold.features.predict_3Di:get_embeddings:493 - BUDPYTUS_CDS_0221 MZ501093.1 prediction has length 0
```

which I would assume then causes the 0-d array issue. 

When Phold tries to write the 3Di sequences, it will iterate over an empty array for these proteins, which clearly errors out.

I'll put in some fix for the next version. 

To practically solve your error (as it'll take a while before the next release and in any case you'll miss potential annotations for these proteins!), the embedding failure that is the root cause of this error is probably caused by your hardware. Therefore, I'd recommend:

1. Rerunning the command - I feel like this embedding error is a bit random and might disappear next time you run it; or
2. Getting access to a computer with bigger GPU (or a more beefy CPU); or
3. Use the colab notebook (especially if you have a small number of phages) that will give you a GPU with enough memory https://colab.research.google.com/github/gbouras13/phold/blob/main/run_pharokka_and_phold_and_phynteny.ipynb
4. Send me the genbank file(s) and I can run it for you, shouldn't be a massive deal :) george.bouras@adelaide.edu.au

George

**Comment by vdruelle** [comment]
Hi @gbouras13,

Thanks a lot for the answer and suggestions to fix this problem. I'll give it a try in the following days.

Have a great day !
Valentin

**Comment by vdruelle** [comment]
Hi @gbouras13,

First of all, sorry for the delay I only recently got time to test this. It took a while but I finally managed to get Phold to run on our HPC that uses A100 GPUs. I'm now using phold 2.0 after a fresh conda env install. Unfortunately it seems I still get the same error:

```
2024-10-14 10:47:11.368 | INFO     | phold.utils.util:begin_phold:72 - phold: annotating phage genomes with protein structures
2024-10-14 10:47:11.368 | INFO     | phold.utils.util:begin_phold:74 - You are using phold version 0.2.0
2024-10-14 10:47:11.369 | INFO     | phold.utils.util:begin_phold:75 - Repository homepage is https://github.com/gbouras13/phold
2024-10-14 10:47:11.369 | INFO     | phold.utils.util:begin_phold:76 - You are running phold run
2024-10-14 10:47:11.369 | INFO     | phold.utils.util:begin_phold:77 - Listing parameters
2024-10-14 10:47:11.369 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --input output/EM11_pharokka/pharokka.gbk
2024-10-14 10:47:11.370 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --output output/EM11_phold
2024-10-14 10:47:11.370 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --threads 8
2024-10-14 10:47:11.370 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --force True
2024-10-14 10:47:11.370 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --prefix phold
2024-10-14 10:47:11.370 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --evalue 0.001
2024-10-14 10:47:11.370 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --database None
2024-10-14 10:47:11.371 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --batch_size 1
2024-10-14 10:47:11.371 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --sensitivity 9.5
2024-10-14 10:47:11.371 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --keep_tmp_files False
2024-10-14 10:47:11.371 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --cpu False
2024-10-14 10:47:11.371 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --omit_probs False
2024-10-14 10:47:11.372 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --finetune False
2024-10-14 10:47:11.372 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --finetune_path None
2024-10-14 10:47:11.372 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --card_vfdb_evalue 1e-10
2024-10-14 10:47:11.372 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --separate False
2024-10-14 10:47:11.372 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --max_seqs 10000
2024-10-14 10:47:11.373 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --save_per_residue_embeddings False
2024-10-14 10:47:11.373 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --save_per_protein_embeddings False
2024-10-14 10:47:11.373 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --only_representatives False
2024-10-14 10:47:11.373 | INFO     | phold.utils.util:begin_phold:79 - Parameter: --ultra_sensitive False
2024-10-14 10:47:11.378 | INFO     | phold.utils.validation:check_dependencies:126 - Foldseek version found is v9.427df8a
2024-10-14 10:47:11.378 | INFO     | phold.utils.validation:check_dependencies:135 - Foldseek version is ok
2024-10-14 10:47:11.378 | INFO     | phold.databases.db:validate_db:361 - Checking Phold database installation in /scicore/home/neher/druell0000/miniconda3/envs/phold_cu124/lib/python3.11/site-packages/phold/database
2024-10-14 10:47:11.461 | INFO     | phold.databases.db:validate_db:364 - All Phold databases files are present
2024-10-14 10:47:11.461 | INFO     | phold.io.handle_genbank:get_genbank:57 - Checking if input output/EM11_pharokka/pharokka.gbk is a Genbank file
2024-10-14 10:47:11.514 | INFO     | phold.utils.validation:validate_input:50 - Successfully parsed input output/EM11_pharokka/pharokka.gbk as a Genbank format file
2024-10-14 10:47:11.664 | INFO     | phold.features.predict_3Di:get_T5_model:121 - Using device: cuda:0
2024-10-14 10:47:11.664 | INFO     | phold.features.predict_3Di:get_T5_model:127 - Loading T5 from: /scicore/home/neher/druell0000/miniconda3/envs/phold_cu124/lib/python3.11/site-packages/phold/database/Rostlab/ProstT5_fp16
2024-10-14 10:47:11.664 | INFO     | phold.features.predict_3Di:get_T5_model:128 - If /scicore/home/neher/druell0000/miniconda3/envs/phold_cu124/lib/python3.11/site-packages/phold/database/Rostlab/ProstT5_fp16 is not found, it will be downloaded
2024-10-14 10:47:30.309 | INFO     | phold.features.predict_3Di:get_T5_model:164 - Rostlab/ProstT5_fp16 loaded
2024-10-14 10:47:30.329 | INFO     | phold.features.predict_3Di:get_embeddings:436 - Beginning ProstT5 predictions
2024-10-14 10:47:30.428 | INFO     | phold.features.predict_3Di:get_embeddings:441 - Using models in half-precision
2024-10-14 10:47:35.797 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0049 MZ501111.1 prediction has length 0
2024-10-14 10:47:38.983 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0165 MZ501111.1 prediction has length 0
2024-10-14 10:47:39.353 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0180 MZ501111.1 prediction has length 0
2024-10-14 10:47:39.599 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0190 MZ501111.1 prediction has length 0
2024-10-14 10:47:39.672 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0193 MZ501111.1 prediction has length 0
2024-10-14 10:47:39.969 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0205 MZ501111.1 prediction has length 0
2024-10-14 10:47:40.237 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0216 MZ501111.1 prediction has length 0
2024-10-14 10:47:40.381 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0222 MZ501111.1 prediction has length 0
2024-10-14 10:47:40.581 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0230 MZ501111.1 prediction has length 0
2024-10-14 10:47:40.731 | WARNING  | phold.features.predict_3Di:get_embeddings:592 - VNNBSPFN_CDS_0236 MZ501111.1 prediction has length 0
...
TypeError: len() of unsized object
```

I get the same error when running phold locally (on CPU). In this case I think this is similar to this issue: https://github.com/gbouras13/phold/issues/74#issue-2584571905

I also tried running this phage genome in the collab notebook and it works there. Digging a bit deeper into the input of Phold (output of Pharokka), I noticed there are big differences there:
- the Pharokka in the google collab uses PHANOTATE_1.5.1 while my phold+pharokka install on the HPC uses PHANOTATE_1.6.5
- looking at the first problematic feature (VNNBSPFN_CDS_0049) the Pharokka files are quite a bit different. The collab file has translation "VITLLNKIPPQNRLKLNLTPTPDLCMIISPTEQRRSSWIILRM" while my install on HPC has translation "H".

Could that be the source of the issue ? I checked the genomes and the beginning of the sequence are the same so I think they are in the same orientation. 
I'm including the files in case you want to check yourself:
[pharokka_collab.txt](https://github.com/user-attachments/files/17362064/pharokka_collab.txt)
[pharokka_HPC.txt](https://github.com/user-attachments/files/17362065/pharokka_HPC.txt)

Thank you for your work and have a great day ! :+1: 
Valentin

**Comment by gbouras13** [comment]
Hi Valentin - that is certainly the issue, pharokka should have phanotate v1.5.1 (the higher versions are broken as you can tell my the “H” cds!)

If you install that version in the pharokka environment on your hpc all should be fine.

george

**Comment by vdruelle** [comment]
Hi again,

Ok great, I'll do that and hopefully things will be fine !
May I suggest adding this information to the README ? Since lots of people run Phold on the output of Pharokka, that seems like an important information.
Alternatively wouldn't it be interesting to include Pharokka in the conda package to make sure the right version is there ?

Best,
Valentin

**Comment by gbouras13** [comment]
Hi @vdruelle  - pharokka's environment specifies that phanotate==1.5.1 be used (well below 1.6.3), so it must be a specific issue for you with the way you've installed it on HPC.  https://github.com/gbouras13/pharokka/blob/master/environment.yml. I would assume you did a source or pip install, not conda, or else 1.5.1 would have been installed automatically. You could also just try `-g prodigal-gv` to use prodigal-gv instead of phanotate

In terms of installing it as a dependency explicitly, pharokka already has a lot of dependencies - hence why I suggest making separate conda envs for pharokka and phold in the documentation - I don't want phold to become a fragile leaning tower of dependencies.

George

**Comment by vdruelle** [comment]
Hi @gbouras13,

That makes sense, seem like a sensible approach. 
Regarding the installation I made them using conda, both locally and on the cluster. It seems I ended up with too recent versions of phanotate in both cases. Maybe there is something wrong there ? On both cases I tried installing Pharokka in the same environment as Phold though. 

From what I see in the `environment.yml` file, the version of Phanotate is `- phanotate >=1.5.0,<1.6.4`, so maybe it went for the most recent version within this limit (the 1.6.4) ?

Thanks a lot for your help and have a great day ! :+1:  
Valentin
